File: LeetCode-C++/LeetCode-C++/二叉树/_102_二叉树的层次遍历.cpp
//
//  _102_二叉树的层次遍历.cpp
//  LeetCode-C++
//
//  Created by warden on 2019/12/11.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/binary-tree-level-order-traversal/

#include <stdio.h>
#include <vector>
#include <queue>
#include "TreeNode.hpp"

using namespace std;

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int>> result;
        queue<TreeNode*> nodes;
        if (root) {
            nodes.push(root);
        }
        size_t size = nodes.size();
        vector<int> curLevel;
        while (!nodes.empty()) {
            if (size == 0) {
                size = nodes.size();
                result.push_back(curLevel);
                curLevel.clear();
            }
            TreeNode* p = nodes.front();
            nodes.pop();
            curLevel.push_back(p->val);
            if (p->left) {
                nodes.push(p->left);
            }
            if (p->right) {
                nodes.push(p->right);
            }
            size--;
        }
        if (curLevel.size() > 0) {
            result.push_back(curLevel);
        }
        return result;
    }
};

/**
     3
    / \
   9  20
     /  \
    15   7
 返回其层次遍历结果：

 [
   [3],
   [9,20],
   [15,7]
 ]
 */
/**
int main(int argc, const char * argv[]) {
    TreeNode* root = generateTreeByLevel({3, 9, 20, INT_MAX, INT_MAX, 15, 7});
    Solution* solution = new Solution();
    vector<vector<int>> result = solution->levelOrder(root);
    for (auto v:result) {
        printVector(&v);
    }
    return 0;
}
 */

File: LeetCode-Swift/LeetCode-Swift/二叉树/_102_二叉树的层次遍历.swift
//
//  _102_二叉树的层次遍历.swift
//  LeetCode-Swift
//
//  Created by warden on 2019/12/11.
//  Copyright © 2019 warden. All rights reserved.
//
// https://leetcode-cn.com/problems/binary-tree-level-order-traversal/

import Foundation

class _102_二叉树的层次遍历 {
    func levelOrder(_ root: TreeNode?) -> [[Int]] {
        var result = [[Int]]()
        if root == nil {
            return result
        }
        var nodes = [root!]
        var size = nodes.count
        var level = [Int]()
        while !nodes.isEmpty {
            if size == 0 {
                size = nodes.count
                result.append(level)
                level.removeAll()
            }
            let p = nodes.removeFirst()
            level.append(p.val)
            if p.left != nil {
                nodes.append(p.left!)
            }
            if p.right != nil {
                nodes.append(p.right!)
            }
            size -= 1
        }
        if level.count > 0 {
            result.append(level)
        }
        return result
    }
}

/**
 给定二叉树: [3,9,20,null,null,15,7],

    3
   / \
  9  20
    /  \
   15   7
返回其层次遍历结果：

[
  [3],
  [9,20],
  [15,7]
]
 */
func test_102() {
    let root = generateTreeByLevel([3, 9, 20, nil, nil, 15, 7])
    let solution = _102_二叉树的层次遍历()
    print(solution.levelOrder(root))
}

